### PR TITLE
Introduce machineConfigServer and nodePool versions

### PR DESCRIPTION
**Pull request by enxebre**
In order to have fine grained control over the versioning of nodePools, machines and nodes we want to decouple the machine config server, the ign endpoint and the user data secret giving the ignition endpoint from the controlPlane lifecycle. This will let us have nodePools coexisting at different versions and orchestrate rolling upgrades from an authoritative source of truth.

To that end this PR introduces a new CRD machineConfigServer and introduce the concept of versioning for nodePools. A nodePool always targets a release version and generates the appropriate mcs for it to get the right ignition config.

```
Albertos-MacBook-Pro:hypershift@albertogarla $ kubectl get mcs -nexample
NAME                 VERSION   HOST
example-extended-a   4.7.3     xxxxx.com
Albertos-MacBook-Pro:hypershift@albertogarla $
```

```
Albertos-MacBook-Pro:hypershift@albertogarla $ kubectl get nodepool -nclusters
NAME                 CLUSTER   NODECOUNT   AUTOSCALING   VERSION   UPGRADING
example-extended-a   example   1           False         4.7.3     False
Albertos-MacBook-Pro:hypershift@albertogarla $
```

Supersedes https://github.com/openshift/hypershift/pull/76

Follow ups:
- Sort out pki generation generically as a service for other components including mcss
- Support authentication https://github.com/openshift/hypershift/pull/92

### CODE REVIEW

File: hypershift-operator/controllers/machineconfigserver/assets/assets.go [hunk 1]
```diff
@@ -0,0 +1,26 @@
+package assets
+
+import "embed"
+
+//go:embed machine-config-server/*
```
Comment (ironcladlou) [inline hypershift-operator/controllers/machineconfigserver/assets/assets.go:5]: We already purged templates from this operator, can we stick with the new pattern of defining all our resources in Go like the hosted control plane operator?

File: api/v1alpha1/machineconfigserver_types.go [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+}
+
+// +kubebuilder:object:root=true
+// +kubebuilder:resource:path=machineconfigservers,shortName=mcs;mcss,scope=Namespaced
+// +kubebuilder:storageversion
+// +kubebuilder:subresource:status
+// +kubebuilder:printcolumn:name="Version",type="string",JSONPath=".status.version",description="Version"
+// +kubebuilder:printcolumn:name="Host",type="string",JSONPath=".status.host",description="Host"
+// MachineConfigServer is the Schema for the MachineConfigServers API
+type MachineConfigServer struct {
```
Comment (ironcladlou) [inline api/v1alpha1/machineconfigserver_types.go:37]: Not entirely clear on the purpose of this new type. Is it true there's a 1:1 affinity between a MachineConfigServer and a NodePool for a specific release image value? And is it true the MachienConfigServer is purely a function of a NodePool? Then for example, could the same job be done if the ignition address a property of the NodePool status? Sorry if this is a dumb Q, trying to understand the model

**Review by ironcladlou — CHANGES_REQUESTED** [review]

File: api/v1alpha1/machineconfigserver_types.go (hunk 2 above)
Comment (enxebre) [reply api/v1alpha1/machineconfigserver_types.go:37]: >Is it true there's a 1:1 affinity between a MachineConfigServer and a NodePool for a specific release image value? And is it true the MachineConfigServer is purely a function of a NodePool? Then for example, could the same job be done if the ignition address a property of the NodePool status? Sorry if this is a dumb Q, trying to understand the model

The functionality this PR is enabling is fine grained control over the versioning of nodePools. By having versioning for nodePools we will be able for instance orchestrate upgrades for nodePools in a cluster one by one.
To achieve this we are exposing a new building block machineConfigServer which is used at convenience here to that end but not limited to it

Comment (ironcladlou) [reply api/v1alpha1/machineconfigserver_types.go:37]: Right but what I'm asking is what does the new intermediate type (MachineConfigServer) enable that isn't possible if the ignition server details were embedded in the NodePool status?

Maybe another clarifying question: when would a MachineConfigServer be deleted?

Comment (enxebre) [reply api/v1alpha1/machineconfigserver_types.go:37]: Ah I see your point. The reason to propose having that logic in its own crd is to support the use case where machines are managed externally and not by a nodePool. In such scenarios as a cluster provider I want to create just a MCS so as a cluster admin you can bring your own machine that consumes the right ignition.

Comment (ironcladlou) [reply api/v1alpha1/machineconfigserver_types.go:37]: I think we all agreed this is a useful abstraction, thanks!

**Pull request merged**